### PR TITLE
Fix #18608: Do not swallow errors before basic TopK

### LINKED ISSUE #18608: [SQLancer] Wrong results with join
### What version of Materialize are you using?

v0.50.0-dev (b3a70feb3)

### How did you install Materialize?

Built from source

### What is the issue?

There is a fresh wrong results issue from SQLancer:
```
DROP DATABASE IF EXISTS database0;
CREATE DATABASE database0 ;
\c database0;
CREATE TABLE IF NOT EXISTS t0(c0 FLOAT    );
CREATE TABLE t1(c0 CHAR(317)     DEFAULT (CAST(0.37833587419206882174194106482900679111480712890625 AS VARCHAR(654))));
CREATE TABLE IF NOT EXISTS t2(c0 DECIMAL , c1 integer   );
CREATE TABLE t3(c0 INT );
CREATE TABLE t4(c0 boolean    );
INSERT INTO t4(c0) VALUES(TRUE);
INSERT INTO t1(c0) VALUES('Ja[LPB'), (''), ((('')||(((((((1677746101)&(1677746101)))<<(((1402123780)&(836896798)))))|(((((-1221621620)|(1402123780)))&(((-1485737887)>>(-1221621620)))))))));
INSERT INTO t2(c1, c0) VALUES(-251187149, 0.7499183096237433954200923835742287337779998779296875);
CREATE VIEW v0(c0) AS (SELECT DISTINCT t2.c1 FROM t0, t3, t1, t2, t4 WHERE t4.c0);
INSERT INTO t4(c0) VALUES(NOT ((NOT (((0.42697382035812925238360548974014818668365478515625)<=(0.78326640921410817242076518596149981021881103515625)))) IS UNKNOWN)), (TRUE), (FALSE);
UPDATE t0 SET c0=0.11376621 WHERE (0.11376621) BETWEEN (0.4413229) AND (0.04469857);
INSERT INTO t2(c0) VALUES(0.60333799444918667642667742256890051066875457763671875);
INSERT INTO t2(c0) VALUES(0.53391115661476107323579753938247449696063995361328125), (0.96599097445518788962459666436188854277133941650390625), (0.50874802079664827036964425133191980421543121337890625);
CREATE INDEX t2i0 ON t2(c1 DESC NULLS LAST);
INSERT INTO t3(c0) VALUES(-1221621620);
DELETE FROM t4;
INSERT INTO t2(c0, c1) VALUES(0.798383958841370588999097890336997807025909423828125, 1402123780), (0.60310610645862705592179509039851836860179901123046875, -886808703), (0.96925146076085255675280905052204616367816925048828125, -1006231220);
INSERT INTO t1(c0) VALUES('}~^KC0');
INSERT INTO t2(c1) VALUES(-1221621620), (-2008997878), (409134752);
INSERT INTO t0(c0) VALUES(0.11376621), (0.1963102), (-1.00623123E9), (0.3170005), (CAST(CAST(((')b0B|#')||(0.8653689345608395)) AS INT) AS REAL));
INSERT INTO t0(c0) VALUES(0.84266704), (0.29847664), (0.28999355);
INSERT INTO t2(c1) VALUES(NULL);
INSERT INTO t2(c0) VALUES(0.3973591531507036922477027474087662994861602783203125);
INSERT INTO t0(c0) VALUES(-1.22162163E9);
INSERT INTO t0(c0) VALUES(0.36171648);
INSERT INTO t4(c0) VALUES(TRUE);
INSERT INTO t0(c0) VALUES(0.05187392);
INSERT INTO t0(c0) VALUES(1677746101);
DROP INDEX IF EXISTS t1i0 RESTRICT;
DROP INDEX IF EXISTS v0i2 CASCADE;
COMMIT;
SELECT ALL t0.c0 FROM t0 JOIN t4 ON (t4.c0) NOT IN (NOT (t4.c0)) RIGHT OUTER JOIN (SELECT - (t3.c0) FROM t1, t3) AS sub0 ON t4.c0 RIGHT OUTER JOIN (SELECT ALL (0.47583646) NOT IN (- (t3.c0)), CAST(t4.c0 AS VARCHAR(813))!~*t1.c0 FROM t3, t4, t2, v0, t0, t1 WHERE NOT (((t4.c0)<(t4.c0))) LIMIT 5633455955921112428) AS sub1 ON t4.c0 WHERE NOT (NOT (t4.c0));-- count: 81536
SELECT SUM(count) FROM (SELECT (NOT (NOT (t4.c0)))::INT as count FROM t0 JOIN t4 ON (t4.c0) NOT IN (NOT (t4.c0)) RIGHT OUTER JOIN (SELECT - (t3.c0) FROM t1, t3) AS sub0 ON t4.c0 RIGHT OUTER JOIN (SELECT (0.47583646) NOT IN (- (t3.c0)), CAST(t4.c0 AS VARCHAR(813))!~*t1.c0 FROM t3, t4, t2, v0, t0, t1 WHERE NOT (((t4.c0)<(t4.c0))) LIMIT 5633455955921112428) AS sub1 ON t4.c0) as res;-- count: 61152;
```
This is using the NoREC check, so we expect the final count to be the same as the number of results in the query before.

The issue can be reproduced, I haven't investigated the root cause further.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by tokenrove**
Commit 2ca9e0979e36faf8df3b6ba403996109c874dff3 added error validation to non-monotonic Top-K; unfortunately, we replaced rather than concatenated the error stream, which means that errors that happen earlier were suppressed, as demonstrated in #18608.

After this, I did double-check all the other assignments to `err_collection` in the render code just to make sure...

### Motivation

 - This PR fixes a recognized bug: #18608.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by vmarcos** [comment]
We might have an issue also here: https://github.com/MaterializeInc/materialize/blob/b57e2f40771da6a7f8eeebed2beb78e1163da74f/src/compute/src/render/reduce.rs#L64-L68

**Comment by tokenrove** [comment]
> We might have an issue also here:
> 
> https://github.com/MaterializeInc/materialize/blob/b57e2f40771da6a7f8eeebed2beb78e1163da74f/src/compute/src/render/reduce.rs#L64-L68

Oof, sure enough.

**Review by vmarcos — APPROVED** [review]
LGTM!

**Comment by aalexandrov** [comment]
@benesch This fixes a correctness issue, so I would nominate it for back-porting the fix to `0.49.x`.

**Review by def- — APPROVED** [review]
Thanks for fixing this super quickly and adding the test!

File: test/testdrive/github-18608.td [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
```
Comment (def-) [inline test/testdrive/github-18608.td:1]: This could probably be SLT too, right?

Comment (tokenrove) [reply test/testdrive/github-18608.td:1]: I could make this an SLT instead if you'd prefer.

**Comment by benesch** [comment]
> @benesch This fixes a correctness issue, so I would nominate it for back-porting the fix to `0.49.x`.

Just in case a customer runs into it? That seems reasonable to me, to have it on hand, though we can't proactively roll it out to folks because the maintenance window for v0.49 has already passed.

If someone would like to lead the charge and backport this into v0.49.x just to have it on hand, I'm on board, though I think it's also fine to wait until someone runs into this and backport on demand. v0.50 will ship soon.

**Pull request merged**